codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -436,6 +423,62 @@ public void operationComplete(ChannelFuture channelFuture) {
         }
     }
 
+    private void writePushPromise(final ChannelHandlerContext ctx, Http2PushPromiseFrame pushPromiseFrame,
+                                  final ChannelPromise promise) {
+        if (isStreamIdValid(pushPromiseFrame.pushStream().id())) {
+            encoder().writePushPromise(ctx, pushPromiseFrame.stream().id(), pushPromiseFrame.pushStream().id(),
+                    pushPromiseFrame.http2Headers(), pushPromiseFrame.padding(), promise);
+        } else {
+            initializeNewStream(ctx, (DefaultHttp2FrameStream) pushPromiseFrame.pushStream(), promise);
+            final int streamId = pushPromiseFrame.stream().id();
+
+            encoder().writePushPromise(ctx, streamId, pushPromiseFrame.pushStream().id(),
+                    pushPromiseFrame.http2Headers(), pushPromiseFrame.padding(), promise);
+
+            if (promise.isDone()) {
+                handleHeaderFuture(promise, streamId);
+            } else {
+                numBufferedStreams++;
+                // Clean up the stream being initialized if writing the headers fails and also
+                // decrement the number of buffered streams.
+                promise.addListener(new ChannelFutureListener() {
+                    @Override
+                    public void operationComplete(ChannelFuture channelFuture) {
+                        numBufferedStreams--;
+
+                        handleHeaderFuture(channelFuture, streamId);
+                    }
+                });
+            }
+        }
+    }
+
+    public void initializeNewStream(ChannelHandlerContext ctx, DefaultHttp2FrameStream http2FrameStream,
```

Comment:
@hyperxpro why is this public ?

Reply:
Also this should return `boolean` so when using this you can detect when it failed and when not to early return.

Reply:
Some use case needs Stream ID to be generated early before sending HeadersFrame. That's why I made it public. If this approach is not good then I will change it back to package-private.

My use case:
```java
Http2FrameCodec.DefaultHttp2FrameStream http2FrameStream = (Http2FrameCodec.DefaultHttp2FrameStream) newStream();
frameCodec.initializeNewStream(ctx, http2FrameStream, promise);
long id = httpFrame.id();

// Put the stream ID and Outbound Property into the map.
addStream(new OutboundProperty(id, http2FrameStream, httpFrame.protocol()));
```

Reply:
I think we need to solve this in a better way...  Maybe we should at least expose this via the `Http2ChannelDuplexHandler`